### PR TITLE
Range manipulator

### PR DESCRIPTION
**Pull request by agirault**
- Add `MouseManipulator` to hold mouse and modifier keys properties, and extend it with `CameraManipulator` for manipulators using the camera.
- Add `RangeManipulator` to generate generic values between a range with a specific step.
- Add example for `RangeManipulator`:
  - control window vertical grab
  - control level with horizontal grab
  - control Z slice with scroll/pinch

### CODE REVIEW

**Comment by agirault** [comment]
@jourdain @floryst

File: Sources/Interaction/Manipulators/CameraManipulator/index.js [hunk 1]
```diff
@@ -33,38 +26,21 @@ function vtkCameraManipulator(publicAPI, model) {
...
 
 // ----------------------------------------------------------------------------
 
 export function extend(publicAPI, model, initialValues = {}) {
   Object.assign(model, DEFAULT_VALUES, initialValues);
 
-  // Object methods
-  macro.obj(publicAPI, model);
+  // Inheritance
+  vtkMouseManipulator.extend(publicAPI, model, initialValues);
```
Comment (agirault) [inline Sources/Interaction/Manipulators/CameraManipulator/index.js]: @jourdain can you confirm this part is right?

File: Sources/Interaction/Style/InteractorStyleManipulator/index.js [hunk 2]
```diff
@@ -288,8 +288,12 @@ function vtkInteractorStyleManipulator(publicAPI, model) {
...
     if (model.currentManipulator) {
       publicAPI.invokeStartInteractionEvent({ type: 'StartInteractionEvent' });
-      model.currentManipulator.setCenter(model.centerOfRotation);
-      model.currentManipulator.setRotationFactor(model.rotationFactor);
+      if (typeof model.currentManipulator.setCenter === 'function') {
+        model.currentManipulator.setCenter(model.centerOfRotation);
+      }
+      if (typeof model.currentManipulator.setRotationFactor === 'function') {
+        model.currentManipulator.setRotationFactor(model.rotationFactor);
+      }
```
Comment (agirault) [inline Sources/Interaction/Style/InteractorStyleManipulator/index.js]: @jourdain any better way to do this? Could check that `currentManipulator isinstanceof vtkCameraManipulator`. instead, but requires importing `vtkCameraManipulator`.

**Comment by jourdain** [comment]
I was thinking to have it that way

```
publicAPI.setHorizontalListener = (min, max, value, step, callback) => {
  model.horizontalListener = { min, max, value, step, callback };
  publicAPI.modified();
}

publicAPI.removeHorizontalListener = () => {
  if (model.horizontalListener) {
    delete model.horizontalListener;
    publicAPI.modified();
  }
}
```

File: Sources/Interaction/Style/InteractorStyleManipulator/index.js (hunk 2 above)
Comment (jourdain) [reply Sources/Interaction/Style/InteractorStyleManipulator/index.js]:
```
model.currentManipulator.set(publicAPI.get('centerOfRotation', 'rotationFactor'), true);
```

Comment (jourdain) [reply Sources/Interaction/Style/InteractorStyleManipulator/index.js]: You will have to rename center to be centerOfRotation on the various manipulator and properly use it that way.

File: Sources/Interaction/Manipulators/CameraManipulator/index.js (hunk 1 above)
Comment (jourdain) [reply Sources/Interaction/Manipulators/CameraManipulator/index.js]: yes it is

File: Sources/Interaction/Manipulators/RangeManipulator/index.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    const lastPos = interactor.getLastAnimationEventPosition(lastPtr);
+
+    if (!pos || !lastPos || !renderer) {
+      return;
+    }
+
+    // Scale by viewport size
+    const size = interactor.getView().getViewportSize(renderer);
+
+    if (model.horizontalListener !== null) {
```
Comment (jourdain) [inline Sources/Interaction/Manipulators/RangeManipulator/index.js]: `if (model.horizontalListener) {`

File: Sources/Interaction/Manipulators/RangeManipulator/index.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      const dx = (pos.x - lastPos.x) / size[0];
+      processDelta(model.horizontalListener, dx);
+    }
+    if (model.verticalListener !== null) {
```
Comment (jourdain) [inline Sources/Interaction/Manipulators/RangeManipulator/index.js]: `if (model.verticalListener){`

File: Sources/Interaction/Manipulators/RangeManipulator/index.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      const dy = (pos.y - lastPos.y) / size[1];
+      processDelta(model.verticalListener, dy);
+    }
+  };
+
+  //-------------------------------------------------------------------------
+  publicAPI.onPinch = (interactor) => {
+    if (model.scrollListener === null) {
```
Comment (jourdain) [inline Sources/Interaction/Manipulators/RangeManipulator/index.js]: if (!model.scrollListener) {

File: Sources/Interaction/Style/InteractorStyleManipulator/index.js (hunk 2 above)
Comment (jourdain) [reply Sources/Interaction/Style/InteractorStyleManipulator/index.js]: If you don't want to do it. Then do that instead...

```
if (model.currentManipulator.setRotationFactor) {
         model.currentManipulator.setRotationFactor(model.rotationFactor);
 }
```

**Pull request merged**